### PR TITLE
 Fix the flaky test testDoDownloadDirectory_PathTraversalProtection

### PR DESCRIPTION
**Pull request by sharatchandrag**
**Summary**
Add path-traversal protection to [doDownloadDirectory](cci:1://file:///Users/sharatchandra.g/IdeaProjects/multicloudj-forked/multicloudj/blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java:933:2-1030:3) and re-enable disabled test

**Description**
- Add path-traversal protection in `GcpBlobStore#doDownloadDirectory`: each blob's post-strip destination is resolved against a canonicalized `safeTargetDir = targetDir.toAbsolutePath().normalize()` and rejected if it escapes via `Path#startsWith` (segment-based, not lexical).
- Rejected blobs are reported as `FailedBlobDownload` entries with a `SecurityException("Blocked path traversal for blob: <name>")` and are filtered before `TransferManager.downloadBlobs(...)`, so attacker-controlled keys never reach the GCS transfer pipeline.
- Folder-marker handling and the early-return on empty `blobInfos` are preserved; `stripPrefix` semantics continue to match `TransferManager.setStripPrefix`.
- Re-enable `GcpBlobStoreTest#testDoDownloadDirectory_PathTraversalProtection` (was `@Disabled("Failing because of temp directory permission")`). The original test was unfixable as written — it verified `Blob#downloadTo` (never called by the implementation) and did not stub `TransferManager`. Rewritten to assert the real contract: malicious blob is in `failedTransfers` with a `SecurityException`, destination is outside `tempDir`, and `mockTransferManager.downloadBlobs(...)` is never invoked.
- No public API change; legitimate keys (including ones with `..` segments that stay within the target directory) are unaffected. [GcpBlobStoreTest](cci:2://file:///Users/sharatchandra.g/IdeaProjects/multicloudj-forked/multicloudj/blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java:130:0-4050:1) passes (300/300).

### CODE REVIEW

**Review by iamabhilaksh — COMMENTED** [review]
Thanks for putting this together!
Could you shed a little more light on the context behind this? It would be really helpful to understand the exact root cause, how this specific implementation fixes it, and the steps you took to verify it. Thanks!

**Comment by sharatchandrag** [comment]
@iamabhilaksh  Please note:
**Context:** **(Aws and Ali sdk handles this but in case of Gcp we need to handle this)**
When downloading a directory of blobs, we need to check where each file would land. A blob with a malicious key like prefix/../../../etc/passwd could escape the target folder and write files to unintended locations on the filesystem.

How the fix works: Before downloading each blob, we now check — "does this file's destination path actually live inside the intended target folder?" If not, we skip it and record it as a failed download with a security error. 

Example:
Target folder: /tmp/downloads
Malicious blob key: prefix/../../../etc/passwd → resolves to  the root directory /etc/passwd → :x: blocked
Normal blob key: prefix/subdir/file.txt → resolves to /tmp/downloads/subdir/file.txt → :white_check_mark: allowed

**Review by SriramGuduri — APPROVED** [review]
LGTM

File: blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java
Comment (iamabhilaksh) [inline blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java:1]: **Suggestion:** There's no test for the `prefix=null` path. When `prefix` is null, `stripPrefix`
  is `""`, so `relative = name` (the raw blob key), and `destination = safeTargetDir.resolve(name)
  .normalize()`. A blob named `"../../../etc/passwd"` should still be blocked. Worth adding a
  test to confirm the protection holds when no prefix is specified?

Comment (iamabhilaksh) [inline blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java:1]: **Suggestion:** The test only covers the all-blocked case (single malicious blob, blobInfos
  ends up empty). There's no test for the mixed case — one malicious + one legitimate blob —
  where we'd want to assert that `downloadBlobs` IS invoked with exactly the legitimate blob
  while the malicious one stays in `failedTransfers`. That's the load-bearing scenario:
  path-traversal protection working alongside normal downloads. Could we add a second test
  case for that path?

File: blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java [hunk 1]
```diff
@@ -1020,19 +1020,41 @@ protected DirectoryDownloadResponse doDownloadDirectory(DirectoryDownloadRequest
...
-          blobInfos.add(blob);
+        if (isFolderMarker(blob)) {
+          continue;
+        }
+        // Path-traversal protection: ensure the destination computed from the
+        // (possibly attacker-controlled) blob name stays under safeTargetDir.
+        String name = blob.getName();
+        String relative =
+            name.startsWith(stripPrefix) ? name.substring(stripPrefix.length()) : name;
+        Path destination = safeTargetDir.resolve(relative).normalize();
```
Comment (iamabhilaksh) [inline blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java:1044]: **Question:** `Path#startsWith` is a segment-based check, not a lexical prefix match, so it
  correctly blocks `safeTargetDir/../sibling` after `normalize()` resolves it. One edge case:
  if `safeTargetDir` itself normalizes to `/` on an unusual mount (e.g. a container root), then
  every path starts with it and the check becomes a no-op. Is that worth a comment, or do we
  consider that out-of-scope for this fix?

File: blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java [hunk 2]
```diff
@@ -1020,19 +1020,41 @@ protected DirectoryDownloadResponse doDownloadDirectory(DirectoryDownloadRequest
               Storage.BlobField.SIZE,
               Storage.BlobField.GENERATION));
 
+      // Resolve the canonical target directory once so path-traversal checks below
+      // compare absolute, normalized paths (e.g. blocks "../../../etc/passwd").
+      final Path safeTargetDir = targetDir.toAbsolutePath().normalize();
+      final String stripPrefix = prefix != null ? prefix : "";
```
Comment (iamabhilaksh) [inline blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java:1026]: nit : Use StringUtils.EMPTY instead.

```
final String stripPrefix = prefix != null ? prefix : StringUtils.EMPTY;
```

**Review by iamabhilaksh — COMMENTED** [review]
Great work! The path-traversal protection is secure and properly implemented.

I have one minor nit regarding consistency:

In the traversal check, you compute the failed destination using safeTargetDir (which is absolute). However, in the `TransferManager` failure-handling block later on, you use targetDir (*which could be relative, depending on user input*).

`Suggestion:` Use safeTargetDir in both places so API consumers receive a consistent, absolute path format for all FailedBlobDownload errors.

Diff
```
- Path destination = targetDir.resolve(relative).normalize();
+ Path destination = safeTargetDir.resolve(relative).normalize();
```

Other than that, this looks excellent! 🚀

**Review by iamabhilaksh — APPROVED** [review]
Left some minor comments. Rest LGTM 👍

File: blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java (hunk 1 above)
Comment (sharatchandrag) [reply blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java:1044]: This is out of scope usecase for us, its unsual for the users to pass the root directory as the download directory and also most systems will reject such downloads

File: blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java
Comment (sharatchandrag) [reply blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java:1]: Yes this is good point, will take up this in the next iteration

Comment (sharatchandrag) [reply blob/blob-gcp/src/test/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStoreTest.java:1]: Yes, will take this edge case in the next iteration

**Review by SriramGuduri — APPROVED** [review]
LGTM — clean, correct path-traversal protection using the resolve-normalize-startsWith pattern. Malicious blobs are filtered before reaching TransferManager.

**Nitpicks (non-blocking):**

1. `GcpBlobStoreTest.java` — consider using `assertInstanceOf(SecurityException.class, failure.getException())` (JUnit 5.8+) instead of `assertTrue(x instanceof Y)` for clearer failure messages.

2. `GcpBlobStore.java` — minor: `stripPrefix` and the path-traversal check both derive from the same request field. Just a note to keep these in sync if the prefix semantics ever diverge.

File: blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java [hunk 3]
```diff
@@ -1020,19 +1020,41 @@ protected DirectoryDownloadResponse doDownloadDirectory(DirectoryDownloadRequest
...
-        if (!isFolderMarker(blob)) {
-          blobInfos.add(blob);
+        if (isFolderMarker(blob)) {
+          continue;
+        }
+        // Path-traversal protection: ensure the destination computed from the
+        // (possibly attacker-controlled) blob name stays under safeTargetDir.
+        String name = blob.getName();
+        String relative =
+            name.startsWith(stripPrefix) ? name.substring(stripPrefix.length()) : name;
```
Comment (LihaoLiuXs) [inline blob/blob-gcp/src/main/java/com/salesforce/multicloudj/blob/gcp/GcpBlobStore.java:1043]: **The prefix strip here does not match how the GCS SDK computes the destination, so the guard validates a path the SDK will not use — and it rejects legitimate blobs.**

In google-cloud-storage 2.62.0 (the version pinned in `blob-gcp/pom.xml`), `TransferManagerUtils.createDestPath` computes the destination as
`config.getDownloadDirectory().resolve(blob.getName().replaceFirst(config.getStripPrefix(), ""))`.
`String#replaceFirst` treats the strip prefix as a **regex**, while line 1043 strips it as a literal prefix. The two agree only when the prefix contains no regex metacharacters.

Concrete case — `prefixToDownload = "backup(1)/"`, blob key `backup(1)/../evil.txt`:
- SDK: `replaceFirst("backup(1)/", "")` matches nothing (that pattern means `backup1/`), so it writes `<dir>/backup(1)/../evil.txt` → `<dir>/evil.txt`, safely inside the target directory.
- This code: strips literally → `../evil.txt` → resolves to `<parent>/evil.txt` → the blob is failed with `SecurityException`.

A download that works today starts failing, which contradicts "legitimate keys (including ones with `..` segments that stay within the target directory) are unaffected". The same mismatch makes the reported `destination` wrong.

Suggested fix: make both sides strip identically, e.g. `setStripPrefix(Pattern.quote(prefix))` at line 1067 — `getStripPrefix()` is consumed only by `createDestPath`, and quoting also avoids `PatternSyntaxException` for prefixes such as `data[2024/`.
